lib/cache/plugins.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+type pluginIndex string
+
+const pluginNameIndex pluginIndex = "name"
+
+func newPluginsCollection(service services.Plugins, watch types.WatchKind) (*collection[types.Plugin, pluginIndex], error) {
+	if service == nil {
+		return nil, trace.BadParameter("missing parameter Plugin Service")
+	}
+
+	return &collection[types.Plugin, pluginIndex]{
+		store: newStore(
+			types.KindPlugin,
+			types.Plugin.Clone,
+			map[pluginIndex]func(types.Plugin) string{
+				pluginNameIndex: types.Plugin.GetName,
+			},
+		),
+		fetcher: func(ctx context.Context, loadSecrets bool) ([]types.Plugin, error) {
+			var startKey string
+			var allItems []types.Plugin
+
+			for {
+				items, nextKey, err := service.ListPlugins(ctx, apidefaults.DefaultChunkSize, startKey, loadSecrets)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+				if !loadSecrets {
+					for i, plugin := range items {
+						items[i] = plugin.WithoutSecrets().(types.Plugin)
+					}
+				}
+				allItems = append(allItems, items...)
+				if nextKey == "" {
+					break
+				}
+				startKey = nextKey
+			}
+			return allItems, nil
+		},
+		headerTransform: func(hdr *types.ResourceHeader) types.Plugin {
+			return &types.PluginV1{
+				Kind:    hdr.Kind,
+				Version: hdr.Version,
+				Metadata: types.Metadata{
+					Name: hdr.Metadata.Name,
+				},
+			}
+		},
+		watch: watch,
+	}, nil
+}
+
+// GetPlugin retrieves a plugin by name from the cache or fallback backend.
+func (c *Cache) GetPlugin(ctx context.Context, name string, withSecrets bool) (types.Plugin, error) {
+	_, span := c.Tracer.Start(ctx, "cache/GetPlugin")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.plugins)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		item, err := c.Config.Plugin.GetPlugin(ctx, name, withSecrets)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return stripPluginSecrets(item, withSecrets), nil
+	}
+
+	plugin, err := rg.store.get(pluginNameIndex, name)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return plugin, nil
+}
+
+// GetPlugins returns all plugins from the cache or backend.
+func (c *Cache) GetPlugins(ctx context.Context, withSecrets bool) ([]types.Plugin, error) {
+	_, span := c.Tracer.Start(ctx, "cache/GetPlugins")
+	defer span.End()
+
+	rg, err := acquireReadGuard(c, c.collections.plugins)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	if !rg.ReadCache() {
+		plugins, err := c.Config.Plugin.GetPlugins(ctx, withSecrets)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if !withSecrets {
+			for i, plugin := range plugins {
+				plugins[i] = plugin.WithoutSecrets().(types.Plugin)
+			}
+		}
+	}
+
+	var plugins []types.Plugin
+	for item := range rg.store.resources(pluginNameIndex, "", "") {
+		plugins = append(plugins, item.Clone())
```

Comment:
I think we still have to do a `WithoutSecrets()` here if the request does not have `withSecrets`; I'm not sure if we can switch between WithoutSecrets and Clone tho since I would imagine that WithoutSecrets will just return the original item if it doesn't have secrets already.

Reply:
(here and in the other methods)